### DISCUSSION TITLE
Is it possible to minify html

### DISCUSSION THREAD

**Discussion by andriiradkevych** [freshframework/fresh#2213] [General]
I return html and there can be added some `<script>code here</script>` and `<style>code here </style>` , can I compress it and minify somehow ? Or I have to find my own solution ? Maybe you can give advice of library which I can use in `deno` for minify `css` and `js` strings 

<img width="486" alt="Screenshot 2024-01-05 at 16 44 16" src="https://github.com/denoland/fresh/assets/26725494/081e1c6d-102a-4ffe-b45f-1b35065997ce">

Example of returned style , want it to be compressed in response . Same with `<script>`

**Comment by marvinhagemeister** [comment]
Typically HTML minifiers only minify HTML and not other languages like CSS and JavaScript. Fresh already minifies the HTML, but doesn't minify inline stylesheets or scripts at the moment. You can do minify those sections manually by calling into esbuild's transform API before passing it to the JSX.

**Reply by andriiradkevych** [reply]
I can't find solution from deno libraries @marvinhagemeister

**Reply by andriiradkevych** [reply]
Can you give some advice what can I use to minimize `css` and `js` string by deno ?

**Reply by marvinhagemeister** [reply]
Sure, no need for deno-specific libraries. Using plain `esbuild` is enough.

```ts
import { stop, transform } from "https://deno.land/x/esbuild@v0.19.4/mod.js";

const result = await transform(`.foo { color: #ff0000 }`, {
  minify: true,
  sourcemap: false,
  // Set the appropriate loader here. Esbuild supports various types here for TS or JS
  loader: "css",
});

// optional: shutdown esbuild
stop();

console.log("Minified css:", result.code);
// Logs: Minified css: .foo{color:red}
```

**Reply by andriiradkevych** [reply]
Thank you a lot . Hope it will be useful for someone

**Reply by andriiradkevych** [reply]
@marvinhagemeister tried to use `transformSync` , and it return `The "transformSync" API does not work in Deno
` . I need sync method, is there any other ways ?

**Reply by marvinhagemeister** [reply]
Why do you need it to be synchronous?

**Reply by andriiradkevych** [reply]
```
import { ComponentType } from 'preact';

export const withStyles = function <P extends object>(
  component: ComponentType<P>
) {
  return (styles: string) => {
    return function ({ ...props }: P) {
      const Component = component;

      return (
        <>
          <style
            dangerouslySetInnerHTML={{
              __html: styles,
            }}
          />
          <Component {...props} />
        </>
      );
    };
  };
};

```

I have `hoc` , and when I do it async here , it doesn't work . How I use this hoc is `withStyles(Component)(styles)`, where styles is style string 

```
import { ComponentType } from 'preact';

import { transform } from 'minify';

export const withStyles = function <P extends object>(
  component: ComponentType<P>
) {
  return (styles: string) => {
    return async function ({ ...props }: P) {
      const Component = component;

      const minifiedString = await transform(styles, {
        minify: true,
        sourcemap: false,
        loader: 'css',
      });

      return (
        <>
          <style
            dangerouslySetInnerHTML={{
              __html: minifiedString.code,
            }}
          />
          <Component {...props} />
        </>
      );
    };
  };
};
``` 

I mean code like this  @marvinhagemeister  doesn't work

**Reply by marvinhagemeister** [reply]
Async components are not supported in Preact, but you can make the high order function async. With the function signature you provided, the styles don't change on every render. So it would be wasteful to re-compress the same styles over and over again anyway. I think you're looking more for something like this:

```tsx
const withStyles = async <P extends object>(Component: ComponentType<P>, css: string) => {
  const minifed = await transform(css, {
    minify: true,
    sourcemap: false,
    loader: 'css',
  });
  
  const innerHTML = { __html: minified.code }
  
  return (props: P) => {
    return (
      <>
        <style dangerouslySetInnerHTML={innerHTML} />
        <Component {...props} />
      </>
    )
  }
}

// Usage
const MyStyledComponent = await withStyles(Foo, `.foo { color: red }`)
```

If you still want to still partially apply values you can always wrap it again:

```tsx
const withStylesForFoo = async (css: string) => withStyles(Foo, css);
```
